core/blockchain.go

```diff
@@ -1899,6 +1942,12 @@ func (bc *BlockChain) insertSideChain(block *types.Block, it *insertIterator) (i
 	)
 	parent := it.previous()
 	for parent != nil && !bc.HasState(parent.Root) {
+		if bc.StateRecoverable(parent.Root) {
+			if err := bc.triedb.Recover(parent.Root); err != nil {
```

Comment:
Actually, here we already find "external" chain is better than our local then we want to import them as the new canonical chain.

In `Recover`, a few reverts are required upon of the current disk layer, which theoretically can generate a batch of layers and be maintained in the snapDatabase as well. 

But we only want to maintain at most 256 layers in  snapDatabase(otherwise, OOM can be an issue) which means we need to throw away the topmost layers somehow. The corresponding state of *current canonical chain* will still be missing. For sake of simplification, we just keep the new disk layer after Recover.